Add SVG support

danielpuglisi opened the pull request:
This PR adds SVG support (#22).

To support both SVG and PNG during the HTML layout generation I added image data url support: https://github.com/damoiser/qr-bills/compare/master...codegestalt:svg-support?expand=1#diff-1707220acf1a40ad0f84d3991ef6d6f567c0b9b63b390548a0ef7e1047d17989R11

The parameter `qrcode_filepath` has been removed. Using base64 encryption we don't need to touch the file system for the png format anymore and can use the existing layout without having to add additional format logic.

The parameter `qrcode_format` has been added. As when generating HTML with `params[:output_params][:format]` there was no way of specifying the qr code format. `params[:qrcode_format]` solves this. Although I think this is a bit smelly. Imo `QRBills.generate` should just focus on generating the entire HTML payslip and remove support for creating `png` / `svg` qr codes and leave that logic to `QRGenerator.create`. I would even go so far as to remove `QRBills.generate` entirely and tell people to use `QRHTMLayout.create` for generating an entire payslip and `QRGenerator.create` for just qr codes.

danielpuglisi commented:
Added some SVG optimizations based on:

- https://codepen.io/tigt/post/optimizing-svgs-in-data-uris
- https://github.com/rails/sprockets/blob/master/lib/sprockets/context.rb#L277-L303
- https://github.com/rails/sprockets/issues/425

damoiser commented:
thx for the PR! I would need to do some more deep checks on the svg changes, in the meantime, can you please check my feedback and fix the merge conflicts?

danielpuglisi commented:
Merge conflicts are fixed. I don't see any other feedback regarding this PR. Did you add comments somewhere that I might have missed?

Regarding the code changes: There are some potential backwards incompatibilities if existing users call the following methods directly in their code base:

- `QRGenerator.create` does not accept the entire QRParams hash anymore, just the params[:bill_params] as that's the only data needed to create the QR code.
- `QRHTMLLayout.html_layout` now requires a new parameter `qrcode_data_url`.

damoiser commented:
sorry forget to publish the feedbacks, you should see now @danielpuglisi

danielpuglisi commented on qr-bills.gemspec:
```diff
@@ -19,7 +20,9 @@ Gem::Specification.new do |s|
   s.required_ruby_version = ">= 2.7.1"
   s.add_runtime_dependency("i18n", ">= 1.8.3", "< 2")
-  s.add_runtime_dependency("rqrcode", ">= 1.1.2", "< 2")
+  s.add_runtime_dependency("rqrcode")
```
You're right. Will fix.

danielpuglisi commented on lib/qr-bills/qr-generator.rb:
```diff
@@ -1,6 +1,86 @@
+    svg_code.gsub!(/<\/svg>/, '</g>\0')
+    svg_code.gsub!(/<\/g>/, "#{svg_swiss_cross}</g>")
+
+    # Optimize SVG for being URI-escaped
```
It's because of encoding within <img src="">. Single quotes do not need encoding within double quotes which gives a slightly better performance. The SVG comments are not needed as well. For the new lines / whitespace removal I think we can address this directly within the swiss_cross.svg file as rqrcode will already produce a slim svg string.

damoiser replied:
ok thx!

danielpuglisi commented on lib/qr-bills/qr-generator.rb:
```diff
@@ -35,59 +115,39 @@ module QRGenerator
   #  "//S1/10/10201409/11/181105/40/0:30\r\n" +   # bill information coded for automated booking of payment, data is not forwarded with the payment
   #  "eBill/B/41010560425610173";                 # alternative scheme paramaters, max 100 chars
- 
-  def self.create(params, qrcode_path)
```
Alright, will do that 👍

danielpuglisi commented on README.md:
```diff
@@ -90,7 +93,7 @@ params = QRBills.get_qr_params
 
 # fill the params, for example
 params[:bill_type]                                      = QRBills.get_qrbill_with_creditor_reference_type
-params[:qrcode_filepath]                                = "#{Dir.pwd}/tmp/qrcode-html.png"
```
Will do that 👍

danielpuglisi commented:
Latest commit addresses your comments: 

- Added version constraint to rqrcode gem
- Optimized swiss_cross.svg and SVG generation
- Updated methods to be backwards compatible
- Added a deprecation warning for the png variant that touches the filesystem

damoiser commented:
thx a lot for the great PR @danielpuglisi !

danielpuglisi commented:
You're welcome, thx for merging 🙏😊
